Add new functionality required by 6-sprint

AleksMosYanPrac opened the pull request:
feat: add new functionality required by 6-sprint

feat: add new option for history manager
test: add possibility to configure TaskManager
feat: add "delete from history" functionality
refactor: apply "delete from history" functionality
refactor: add implementation of required HistoryRepository
refactor: rename implementation of HistoryRepositories
fix: add null-validation when delete subtask without epic
style: correct names of methods
test: refactor tests for different history repository
test: refactor tests
test: refactor tests with TestDataDTO
test: add tests for new functionality
test: add TestDataDTO class with test DTO data
test: add tests to new history repository
test: add test for "delete from history" functionality

AleksMosYanPrac committed:
feat: add new functionality required by 6-sprint

feat: add new option for history manager
test: add possibility to configure TaskManager
feat: add "delete from history" functionality
refactor: apply "delete from history" functionality
refactor: add implementation of required HistoryRepository
refactor: rename implementation of HistoryRepositories
fix: add null-validation when delete subtask without epic
style: correct names of methods
test: refactor tests for different history repository
test: refactor tests
test: refactor tests with TestDataDTO
test: add tests for new functionality
test: add TestDataDTO class with test DTO data
test: add tests to new history repository
test: add test for "delete from history" functionality

EugeneLenkevich commented on src/ru/yandex/practicum/kanban/repository/impls/InMemoryHistoryRepositoryWithoutCopyAndLimitElements.java:
```diff
@@ -0,0 +1,125 @@
+    }
+
+    @Override
+    public List<Task> get() {
```
Метод лучше назвать например list(), get() - это обычно про один элемент, немного путает :)

EugeneLenkevich commented on test/ru/yandex/practicum/kanban/repository/InMemoryHistoryRepositoryWithCopyAndLimitElementsTest.java:
```diff
@@ -51,7 +51,19 @@ void canDeleteOldTaskAndPutNewWhenCapacityMoreThan10() {
+
+        historyRepository.deleteFromHistory(taskId);
+
+        Assertions.assertEquals(historySize, historyRepository.get().size());
```
В принципе гут, но надежнее будет еще и явно проверить что этого таска не было в списке. То есть другой разработчик, читающий код, будет видеть что мы проверяем факт что таск удален, то есть ничего додумывать не надо. А вот когда мы только размер проверяем, то мы проверяем удаление элемента косвенно, читающий должен додуматься что же именно мы удаляем. В идеале код должен читаться как текстовое описание бизнес-кейса. в Принципе тесты - это и есть описание бизнес-кейсов.
Проверка размера тут тоже нужна и полезна, но слегка расширить проверки и убрать недосказанность :)

EugeneLenkevich commented on test/ru/yandex/practicum/kanban/repository/InMemoryHistoryRepositoryWithoutCopyAndLimitElementsTest.java:
```diff
@@ -0,0 +1,111 @@
+    @Test
+    void shouldNotSaveRepeatedView() {
+
+        historyRepository.add(TASK_1);
```
А перед этим стоит проверить что список пуст. Так мы убеждаемся что начальные условия заданы правильно. Вдруг после какого то изменения в нашем тесте список будет оставаться с данными между запусками. Тогда тест может работать непредсказуемо и мы будем гадать что произошло :)

EugeneLenkevich commented on src/ru/yandex/practicum/kanban/repository/impls/InMemoryHistoryRepositoryWithoutCopyAndLimitElements.java:
```diff
@@ -0,0 +1,125 @@
+        }
+
+        List<E> toList() {
+            List<E> result = new ArrayList<>();
```
new ArrayList<>(<размер мапы, он же размер списка>);

Список можно создавать сразу нужного размера чтобы избежать возможных накладных расходов на его расширение в случае если элементов много. Мелочь, конечно, но лучше сделать оптимально :)

EugeneLenkevich reviewed (changes requested):
Привет, Александр, спасибо за присланное решение

Отличная работа, тесты особенно порадовали. Что хочется улучшить
 - посмотри, везде ли явно видно по коду теста что он проверяет, то есть тесты должны быть наглядными ведь именно по ним другой разработчик будет разбираться как наш код работает
 - лучше еще проверять начальные условия, тогда наш тест будет падать сразу если что то задано не так, так мы себя обезопасим от взаимного влияния тестов друг на друга при сборке, например. Но посмотри, подумай, возможно не везде, а там где это целесообразно :)

AleksMosYanPrac committed:
refactor: work on comments

style: change access modifier
refactor: add initial size of ArrayList
test: add initial conditions
test: add assertation to "delete from history"
style: rename methods HistoryRepository

EugeneLenkevich commented on src/ru/yandex/practicum/kanban/repository/impls/InMemoryHistoryRepositoryWithCopyAndLimitElements.java:
```diff
@@ -8,27 +8,27 @@
 
     public InMemoryHistoryRepositoryWithCopyAndLimitElements() {
-        this.viewedTask = new ArrayDeque<>();
+        this.viewedTaskList = new ArrayDeque<>();
```
Можно опять таки назвать viewedTasks, то что это список можно увидеть по типу переменой, то есть опять повторяем сами себя :)

ArrayDeque - оригинально, неожиданно, но к месту :)

EugeneLenkevich commented on src/ru/yandex/practicum/kanban/repository/HistoryRepository.java:
```diff
@@ -6,9 +6,9 @@
 public interface HistoryRepository {
 
-    void add(Task task);
+    void addToList(Task task);
```
add было нормально, это глагол "добавить"
list тоже глагол

![image](https://github.com/user-attachments/assets/d3b3ae94-0cd4-46ff-8d31-bb3609a25ee1)

То есть логика тут такая. Методы сервиса - это действия, то есть глаголы, у действий есть дополнительный контекст - это данные на вход и на выход, в нашем случае это параметры метода и возвращаемое значение а так же название самого класса
Разбираем
 - HistoryRepository, то есть он в принципе про историю,  то есть viewed tasks в listOfViewedTasks по сути повторяет второй раз очевидную вещь и вносит неоднозначность тем что дает какое то дополнительное указание, то есть может быть ка кто по другому
- addToList, достаточно add, это означает что добавляем в историю(про имени типа), а тут еще есть указание, что тоже вносит если не однозначность а дополнительные вопросы что это за список,
да и вообще, зачем внешнему коду знать как у нас там внутри хранятся данные - то есть это лишняя информация

EugeneLenkevich reviewed (changes requested):
Привет, спасибо за доработку 

> В программировании есть две сложные проблемы: инвалидация кэша и именование переменных

https://martinfowler.com/bliki/TwoHardThings.html

Кажется немного лишнего переименовал методов, давай вернем назад :)

Момент именно про get, дело в том что у нас есть get-методы, они де геттеры (getters), которые работают именно с одним объектом, потому может путать

https://topjava.ru/blog/gettery-i-settery-v-java

А тесты ок :)

AleksMosYanPrac committed:
refactor: work on comments

refactor: rename fields in HistoryRepository implementations
refactor: rename methods in HistoryRepository

EugeneLenkevich reviewed (approved):
Теперь все отлично, спасибо :)
